returnn/forward.py

```diff
@@ -0,0 +1,213 @@
+        assert os.path.exists(
+            self._model_checkpoint.index_path.get_path()
+        ), "Provided model does not exists: %s" % str(self._model_checkpoint)
+
+    def run(self):
+        # run everything in a TempDir as writing HDFs can cause heavy load
+        with tempfile.TemporaryDirectory(prefix="work_") as d:
+            print("using temp-dir: %s" % d)
+            call = [
+                tk.uncached_path(self.returnn_python_exe),
+                os.path.join(tk.uncached_path(self.returnn_root), "rnn.py"),
+                self.out_returnn_config_file.get_path(),
+            ]
+
+            # stash a possible exception until we finished copying files from the temp work to the actual
+            # work folder to be able to examine the files
+            error = None
+            try:
+                sp.check_call(call, cwd=d)
+            except Exception as e:
+                print("Run crashed - copy temporary work folder")
+                error = e
+
+            # move log and tensorboard
+            shutil.move(os.path.join(d, "returnn.log"), "returnn.log")
+            tensorboard_dirs = glob.glob(os.path.join(d, "eval-*"))
+            for dir in tensorboard_dirs:
+                shutil.move(dir, os.path.basename(dir))
+
+            # move hdf outputs to output folder
+            for k, v in self.out_hdf_files.items():
+                try:
+                    shutil.move(os.path.join(d, k), v.get_path())
+                except Exception as e:
+                    if error is None:
+                        # if we had an error before it is expected to have missing hdf files
+                        # so we do not need to raise anything here then
+                        raise e
+
+            if error:
+                raise error
+
+    @classmethod
+    def create_returnn_config(
+        cls,
+        model_checkpoint,
+        returnn_config,
+        eval_mode,
+        log_verbosity,
+        device,
+        **kwargs,
+    ):
+        """
+
+        :param Checkpoint model_checkpoint:
+        :param ReturnnConfig returnn_config:
+        :param int log_verbosity:
+        :param str device:
+        :param kwargs:
+        :return:
+        """
+        assert device in ["gpu", "cpu"]
+        assert "network" in returnn_config.config
+
+        res = copy.deepcopy(returnn_config)
+
+        config = {"load": model_checkpoint}
+
+        post_config = {
+            "device": device,
+            "log": ["./returnn.log"],
+            "log_verbosity": log_verbosity,
+            "task": "eval" if eval_mode else "forward",
+        }
+
+        if not eval_mode:
+            post_config["forward_override_hdf_output"] = True
+            post_config["output_file"] = "output.hdf"
+
+        config.update(returnn_config.config)
+        post_config.update(returnn_config.post_config)
+
+        res.config = config
+        res.post_config = post_config
+        res.check_consistency()
```

Comment:
maybe it would be more elegant to make use of the new `ReturnnConfig.update` method?

Also be aware about which config overwrites the other. Sometimes we might want the users choice to take precedence (maybe log_verbosity?), sometimes the jobs choice (e.g. Task, output file location, maybe log verbosity) 
Ok, actually nevermind, we want to keep nothing of the users parameters. You should invert the overwriting logic here.

Reply:
This is true. Also in training the user can overwrite important parameters by specifying them in the `ReturnnConfig.config` dict and then get unspecified behavior.

But here I fear the danger that someone in the training phase wants to initialize using an existing model and sets `load = "/path/to/model"` in the training config. Then for forwarding the same training config is used and voila the checkpoint parameter is ignored. Maybe we can also assert that no `"load"` is specified in the given config.